File: homy/integrations/aniworld_downloader/widgets/stats/stats.py
from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import fetch_integration_for_widget as fetch_for_widget

WIDGET = {
    'type': 'aniworld_stats',
    'name': 'AniWorld · Statistics',
    'integration_types': ['aniworld_downloader'],
    'default_size_x': 6,
    'default_size_y': 4,
    'icon': 'bar-chart-2',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
    },
}

CACHE_NS = 'aniworld_stats'
CACHE_TTL = 60


def register(app):
    @app.route('/api/aniworld_stats/data', methods=['GET'], endpoint='widget_aniworld_stats_data')
    def aniworld_stats_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'configured': False, 'message': 'widget_id required'}), 400

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        payload, err, code = fetch_for_widget(widget_id, 'aniworld_downloader', 'v1_stats')
        if err:
            return jsonify({'configured': False, 'message': err}), code

        result = {
            'configured': True,
            'total_downloads': payload.get('total_downloads', 0),
            'completed': payload.get('completed', 0),
            'failed': payload.get('failed', 0),
            'total_episodes': payload.get('total_episodes', 0),
            'last_24h_completed': payload.get('last_24h_completed', 0),
            'average_speed_mbps': payload.get('average_speed_mbps', 0),
            'anime_downloads': payload.get('anime_downloads', 0),
            'series_downloads': payload.get('series_downloads', 0),
            'movie_downloads': payload.get('movie_downloads', 0),
        }
        widget_cache.set(ck, result, ttl=CACHE_TTL)
        return jsonify(result)

File: homy/integrations/aniworld_downloader/widgets/stats/stats.js
window.WidgetRegistry.register('aniworld_stats', {
    _t(key, fallback) {
        if (!window.i18n) return fallback;
        const value = window.i18n.translate(key);
        return value === key ? fallback : value;
    },

    async render(container, widgetData) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/aniworld_stats/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.configured) {
                container.innerHTML = `<div class="muted-text" style="padding:16px;">${data.message}</div>`;
                return;
            }

            const grid = document.createElement('div');
            grid.className = 'aniworld-stats-grid';
            const stats = [
                ['aniworld_stat_total', 'Downloads gesamt', data.total_downloads],
                ['aniworld_stat_completed', 'Abgeschlossen', data.completed],
                ['aniworld_stat_failed', 'Fehlgeschlagen', data.failed],
                ['aniworld_stat_24h', 'Letzte 24h', data.last_24h_completed],
                ['aniworld_stat_speed', 'Ø Geschwindigkeit', `${Number(data.average_speed_mbps || 0).toFixed(1)} Mbit/s`],
                ['aniworld_stat_episodes', 'Episoden gesamt', data.total_episodes],
            ];
            grid.innerHTML = stats.map(([key, label, val]) => `
                <div class="aniworld-stats-card">
                    <div class="num">${val}</div>
                    <div class="lbl">${this._t(key, label)}</div>
                </div>
            `).join('');
            container.appendChild(grid);
        } catch (err) {
            container.innerHTML = `<div class="widget-error">${err.message}</div>`;
        }
    },
});
